### PR TITLE
.Net: Support for YAML and Markdown serializations of semantic functions

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Functions/Functions.UnitTests/Functions.UnitTests.csproj [hunk 1]
```diff
@@ -13,10 +13,17 @@
   <Import Project="$(RepoRoot)/dotnet/src/InternalUtilities/test/TestInternalUtilities.props" />
   <ItemGroup>
     <None Remove="Grpc\protoV3.proto" />
+    <None Remove="Monikers.imagemanifest" />
     <None Remove="OpenAPI\TestPlugins\documentV3_1.yaml" />
     <None Remove="OpenAPI\TestPlugins\nonCompliant_documentV3_0.json" />
     <None Remove="OpenAPI\TestPlugins\openapi.json" />
   </ItemGroup>
+  <ItemGroup>
+    <content Include="Monikers.imagemanifest">
```
Comment (SergeyMenshykh) [inline dotnet/src/Functions/Functions.UnitTests/Functions.UnitTests.csproj:22]: Is it necessary?

File: dotnet/src/Functions/Functions.Markdown/Extensions/KernelFunctionsMarkdownExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    /// Creates an <see cref="ISKFunction"/> instance for a semantic function using the specified markdown text.
+    /// </summary>
+    /// <param name="kernel">Kernel instance</param>
+    /// <param name="text">YAML representation of the <see cref="PromptModel"/> to use to create the semantic function</param>
+    /// <param name="functionName">The function name</param>
+    /// <param name="pluginName">The optional name of the plug-in associated with this method.</param>
+    /// <param name="promptTemplateFactory">>Prompt template factory.</param>
+    /// <param name="loggerFactory"></param>
+    /// <returns>The created <see cref="ISKFunction"/>.</returns>
+    public static ISKFunction FromMarkdown(
```
Comment (RogerBarreto) [inline dotnet/src/Functions/Functions.Markdown/Extensions/KernelFunctionsMarkdownExtensions.cs:46]: What `From` ? 

Although we have the description in the XML, it's not clear the intent of what we are getting from, by default I would assume (this, the Kernel) is what I'm getting from a Markdown, which is not the case.

This only creates an instance or does import it also to the Kernel?
If `Yes`, `ImportFunction` would be my preferred prefix.

Suggestion: Include the Function prefix (Apply for others)
```suggestion
    public static ISKFunction FunctionFromMarkdown(
    //OR
    public static ISKFunction ImportFunctionFromMarkdown(
```

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/SemanticKernel.Abstractions/Models/PromptModel.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+using System.Collections.Generic;
+using System.Text.Json.Serialization;
+using Microsoft.SemanticKernel.AI;
+
+namespace Microsoft.SemanticKernel.Models;
+
+/// <summary>
+/// Prompt model files.
+/// </summary>
+public sealed class PromptModel
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Abstractions/Models/PromptModel.cs:12]: I would change it to have at least `Function` in the name. Not sure if `SemanticFunction` will still be there, if the name is `PromptFunction` then I would suggest
`PromptFunctionModel` or `PromptFunctionConfig` (some variables use config in the name, like semanticFunctionConfig

File: dotnet/src/SemanticKernel.Core/Functions/SKFunction.cs [hunk 4]
```diff
@@ -62,6 +64,22 @@ public static class SKFunction
...
+    /// <summary>
+    /// Creates an <see cref="ISKFunction"/> instance for a semantic function using the specified <see cref="PromptModel"/>.
+    /// </summary>
+    /// <param name="semanticFunctionConfig">Instance of <see cref="PromptModel"/> to use to create the semantic function</param>
+    /// <param name="pluginName">The optional name of the plug-in associated with this method.</param>
+    /// <param name="promptTemplateFactory">>Prompt template factory.</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
+    /// <returns>The created <see cref="ISKFunction"/> wrapper for <paramref name="semanticFunctionConfig"/>.</returns>
+    public static ISKFunction Create(
+        PromptModel semanticFunctionConfig,
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Core/Functions/SKFunction.cs:76]: `Model` or `Config` ?
```suggestion
        PromptFunctionConfig promptFunctionConfig,
```

File: dotnet/src/SemanticKernel.Core/Functions/SemanticFunction.cs [hunk 5]
```diff
@@ -137,6 +108,67 @@ public string ToString(bool writeIndented)
...
+    /// Create a semantic function instance, given a semantic function configuration.
+    /// </summary>
+    /// <param name="pluginName">Name of the plugin to which the function being created belongs.</param>
+    /// <param name="functionName">Name of the function to create.</param>
+    /// <param name="promptTemplateConfig">Prompt template configuration.</param>
+    /// <param name="promptTemplate">Prompt template.</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>SK function instance.</returns>
+    public static ISKFunction FromSemanticConfig(
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Core/Functions/SemanticFunction.cs:121]: If we are not sticking with Semantic Function naming, I would avoid keeping it in any new public API's we create/modify.

- FromPromptConfig 
- FromPromptModel
- FromConfig
- FromModel

File: dotnet/src/SemanticKernel.Core/Functions/SemanticFunction.cs [hunk 6]
```diff
@@ -137,6 +108,67 @@ public string ToString(bool writeIndented)
...
+    /// <summary>
+    /// Create a semantic function instance, given a semantic function configuration.
+    /// </summary>
+    /// <param name="semanticFunctionConfig"></param>
+    /// <param name="pluginName">Name of the plugin to which the function being created belongs.</param>
+    /// <param name="promptTemplateFactory">Prompt template factory</param>
+    /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> to use for logging. If null, no logging will be performed.</param>
+    /// <returns>The created <see cref="ISKFunction"/> wrapper for <paramref name="semanticFunctionConfig"/>.</returns>
+    public static ISKFunction Create(
+        PromptModel semanticFunctionConfig,
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Core/Functions/SemanticFunction.cs:150]:
```suggestion
        PromptFunctionModel config,
        // OR
        PromptFunctionConfig model,
```

File: dotnet/src/SemanticKernel.Abstractions/Models/PromptModel.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+{
+    /// <summary>
+    /// Name of the semantic function.
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Abstractions/Models/PromptModel.cs:15]: Which name are we going forward?
```suggestion
    /// Name of the prompt function.
```

File: dotnet/src/SemanticKernel.Abstractions/Models/PromptFunctionModel.cs [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    /// Name of the kernel function.
+    /// </summary>
+    [JsonPropertyName("name")]
+    public string Name { get; set; } = string.Empty;
+
+    /// <summary>
+    /// Format of the prompt template e.g. f-string, semantic-kernel, handlebars, ...
+    /// </summary>
+    [JsonPropertyName("template_format")]
+    public string TemplateFormat { get; set; } = Microsoft.SemanticKernel.TemplateEngine.PromptTemplateConfig.SemanticKernelTemplateFormat;
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Abstractions/Models/PromptFunctionModel.cs:24]: nit. Avoid long lines (fully qualified names if no conflict exists)

``` 
using Microsoft.SemanticKernel.TemplateEngine;
```

```suggestion
    public string TemplateFormat { get; set; } = PromptTemplateConfig.SemanticKernelTemplateFormat;
```

**Review by RogerBarreto — APPROVED** [review]

**Pull request merged**